hack/generator/pkg/astmodel/reference_graph.go

```diff
@@ -44,27 +44,39 @@ func NewReferenceGraphWithResourcesAsRoots(types Types) ReferenceGraph {
 	return NewReferenceGraph(roots, references)
 }
 
+type ReachableTypes map[TypeName]int // int = depth
+
+func (reachable ReachableTypes) Contains(tn TypeName) bool {
+	_, ok := reachable[tn]
+	return ok
+}
+
 // Connected returns the set of types that are reachable from the roots.
-func (c ReferenceGraph) Connected() TypeNameSet {
+func (c ReferenceGraph) Connected() ReachableTypes {
 	// Make a non-nil set so we don't need to worry about passing it back down.
-	connectedTypes := make(TypeNameSet)
+	connectedTypes := make(ReachableTypes)
 	for node := range c.roots {
-		c.collectTypes(node, connectedTypes)
+		c.collectTypes(0, node, connectedTypes)
 	}
 
 	return connectedTypes
 }
 
 // collectTypes returns the set of types that are reachable from the given
 // 'node' (and places them in 'collected').
-func (c ReferenceGraph) collectTypes(node TypeName, collected TypeNameSet) {
-	if collected.Contains(node) {
+func (c ReferenceGraph) collectTypes(depth int, node TypeName, collected ReachableTypes) {
+	if currentDepth, ok := collected[node]; ok {
 		// We can stop here - we've already visited this node.
+		// But first, see if we are at a lower depth:
+		if depth < currentDepth {
+			collected[node] = depth
```

Comment:
If you lower the depth of `node` then you need to keep processing, and the depths of the things reachable from `node` need to be lowered too.

Reply:
🤔

Reply:
added a test case and fix